src/settings.c

```diff
@@ -100,52 +145,50 @@ void check_and_correct_settings(struct settings *s) {
         }
 }
 
-void load_settings(char *cmdline_config_path)
-{
-
-        FILE *config_file = NULL;
-
-        if (cmdline_config_path) {
-                if (STR_EQ(cmdline_config_path, "-")) {
-                        config_file = stdin;
-                } else {
-                        config_file = fopen(cmdline_config_path, "r");
-                }
+static void apply_settings(FILE *config_file) {
+        if (config_file)
+                load_ini_file(config_file);
 
-                if (!config_file) {
-                        DIE("Cannot find config file: '%s'", cmdline_config_path);
-                }
-        }
-
-        if (!config_file) {
-                config_file = xdg_config("/dunst/dunstrc");
-        }
-
-        if (!config_file) {
-                /* Fall back to just "dunstrc", which was used before 2013-06-23
-                 * (before v0.2). */
-                config_file = xdg_config("/dunstrc");
-        }
-
-        if (!config_file) {
-                LOG_W("No dunstrc found.");
-        }
-
-        load_ini_file(config_file);
         settings_init();
         set_defaults();
```

Comment:
Currently `apply_settings` initializes the settings struct and sets the defaults. If you use apply_settings multiple times, the settings will reset to default before every use. `settings_init` and `set_defaults` should only happen once.

Reply:
I think I got it. I have found my mistake, why I got errors when not calling `set_defaults()` every time. I now have a working solution that only runs `settings_init()` once, which basically only calls `set_defaults()`.

Reply:
> P.S.: OK, here is what my further reading of `options_parser.c` turned up: `set_defaults()` changes `struct settings allowed_settings` but [`save_settings()`](https://github.com/dunst-project/dunst/blob/e973cf8b1c86abf687ab3bd7ea4a7285dafd3018/src/option_parser.c#L542) is what actually applies the settings: [`curr_setting` gets overwritten](https://github.com/dunst-project/dunst/blob/e973cf8b1c86abf687ab3bd7ea4a7285dafd3018/src/option_parser.c#L567), if the entry has a value. So I _think_ it does not hurt, at least, but it even seems like `set_defaults()` _must_ run uncondintionally to get the desired outcome, because I get errors on subsequent calls of `apply_settings()` otherwise.

- `set_defaults()` changes the settings struct and sets them to the internal defaults (see settings_data.h, `setting.default_value`).
- `apply_settings()` changes the settings struct and sets them to the values of the loaded ini file (loaded by `load_ini_file()`). This behaviour should probably be changed. I would suggest to make an ini file struct which contains the data of a single ini file. This can then be passed to `apply_settings()`.
- `settings_init()` sets the settings struct to all 0's. Can't quite remember why, but this was needed. This should only be done once, but currently it doesn't matter if you call this function twice, since there's a safeguard (`count`).

I should probably add some documentation about these functions